File: qyao/198-house-robber/SolutionCircularBuffer.java
@SuppressWarnings({"unused", "UnnecessaryReturnStatement"})
class Solution {
  /*
   * Dynamic programming, the optimal amount for the first k houses is max(S_k-1, S_k-2 + V_k)
   * Using a circular buffer as we only need to remember k-1 and k-2
   */
  public int rob(int[] nums) {
    // fuck you corner cases
    if (nums.length == 0) {
      return 0;
    } else if (nums.length == 1) {
      return nums[0];
    } else if (nums.length == 2) {
      return Math.max(nums[0], nums[1]);
    }

    int[] circularBuffer = new int[2];
    circularBuffer[0] = nums[0];
    circularBuffer[1] = Math.max(nums[0], nums[1]);
    for (int i = 2; i < nums.length; i++) {
      circularBuffer[i % 2] =
          Math.max(circularBuffer[(i - 1) % 2], circularBuffer[(i - 2) % 2] + nums[i]);
    }
    return circularBuffer[(nums.length - 1) % 2];
  }
}

File: qyao/198-house-robber/Solution_circular_buffer.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        """
        Naive:
        let dp[i] be the largest amount that can be obtained from the first i houses, and it can be either including the i-th house or without the i-th house
        then dp[i+1] = max(dp[i-1] + nums[i+1], dp[i])

        Improvement:
        As dp[i] only depends on dp[i-1] and dp[i-2], we only need to remember the most recent 2 instead of everything, reducing the space complexity from O(n) to O(1)
        """
        
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        dp = [0, 0] # initialize circular buffer
        
        dp[0] = 0
        dp[1] = nums[0]
        
        for i in range(2, len(nums)+1):
            dp[i%2] = max(dp[(i-2)%2] + nums[i-1], dp[(i-1)%2])
            
        return dp[len(nums)%2]
        

        
        
